### PR TITLE
Week8

### PR DESCRIPTION
**Pull request by tianahR**
JWT BASICS
Controllers and Routes Setup
Postman Setup
Json Web Token Setup
Authorization header setup 
Auth middleware
Handling Errors
Status codes

Preferred app

### CODE REVIEW

File: 05-JWT-Basics/starter/controllers/main.js [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+//check in the controllers
+
+const login = async(req,res)=>{
+
+    const { username, password } = req.body //// check username, password in post(login) request
+    console.log(username,password)
+
+    //check if username and password exist - if not throw error
+    if (!username || !password) {
```
Comment (jsuoz121) [inline 05-JWT-Basics/starter/controllers/main.js:20]: In a production app, you'd likely build out this condition a bit to check for trailing strings, etc (so we don't create a user " "). Not required for this assignment though!

File: 05-JWT-Basics/starter/controllers/main.js [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    const id = new Date().getDate()
+
+  // try to keep payload small, better experience for user
+  // just for demo, in production use long secret, complex and unguessable string value!!!!!!!!!
+
+   //create new token
+  const token = jwt.sign(
+    { id, username }, //payload
+    process.env.JWT_SECRET, //jwt secret
+    { expiresIn: '30d',} //options
```
Comment (jsuoz121) [inline 05-JWT-Basics/starter/controllers/main.js:37]: I know the instructor hard-codes in the video, but this expired timeline can also be stored as an environment variable or declared as a constant

File: 05-JWT-Basics/starter/errors/custom-error.js [hunk 3]
```diff
@@ -1,7 +1,6 @@
 class CustomAPIError extends Error {
-  constructor(message, statusCode) {
+  constructor(message) {
```
Comment (jsuoz121) [inline 05-JWT-Basics/starter/errors/custom-error.js:2]: Another way you might see this done is the status code is set to INTERNAL SERVER ERROR here and then overridden in the classes that inherit like `bad-request`. 

Neither's necessarily better, just giving another option

File: 05-JWT-Basics/preferred/public/browser-app.js [hunk 4]
```diff
@@ -0,0 +1,71 @@
+const formDOM = document.querySelector('.form')
```
Comment (jsuoz121) [inline 05-JWT-Basics/preferred/public/browser-app.js:1]: Nice for testing!

**Review by jsuoz121 — APPROVED** [review]
Really appreciated how much you did in this assignment - both the preferred and the starter! Nice work

File: 05-JWT-Basics/starter/controllers/main.js (hunk 1 above)
Comment (tianahR) [reply 05-JWT-Basics/starter/controllers/main.js:20]: Okay - thank you. Yes, I think I add more conditions in the preferred assignment but not here :)

File: 05-JWT-Basics/starter/controllers/main.js (hunk 2 above)
Comment (tianahR) [reply 05-JWT-Basics/starter/controllers/main.js:37]: Right - what's the importance of expired timeline  in environment variable? I did that as well on the preferred assignment :)

**Comment by tianahR** [comment]
Thank you very much - I appreciate your feedback !

**Pull request merged**